raysect/core/math/function/float/function1d/interpolate.pyx

```diff
@@ -0,0 +1,538 @@
+        if  interpolation_type not in _INTERPOLATION_TYPES:
+            raise ValueError(f'Interpolation type {interpolation_type} not found.')
+
+        cdef int interpolator_idx = _INTERPOLATION_TYPES.get(interpolation_type)
+
+        extrapolation_type = extrapolation_type.lower()
+        if extrapolation_type not in _EXTRAPOLATION_TYPES:
+            raise ValueError(f'Extrapolation type {extrapolation_type} not found.')
+
+        cdef int extrapolator_idx = _EXTRAPOLATION_TYPES.get(extrapolation_type)
+
+        self.x = np.array(x, dtype=np.float64)
+        self.f = np.array(f, dtype=np.float64)
+        self._x = x
+        self._f = f
+        self._last_index = self._x.shape[0] -1
+        self._extrapolation_range = extrapolation_range
+
+        # dimensions checks
+        if x.ndim != 1:
+            raise ValueError(f'The x array must be 1D. Got {x.shape}.')
+
+        if f.ndim != 1:
+            raise ValueError(f'The x array must be 1D. Got {f.shape}.')
+
+        if x.shape != f.shape:
+            raise ValueError(f'Shape mismatch between x array ({x.shape}) and f array ({f.shape}).')
+
+        # test monotonicity
+        if (np.diff(x) <= 0).any():
+            raise ValueError('The x array must be monotonically increasing.')
+
+        # create extrapolator per extrapolation_type argument
+        if extrapolator_idx == EXT_NONE:
+            self._extrapolator = _ExtrapolatorNone(x, f, extrapolation_range)
+        elif extrapolator_idx == EXT_NEAREST:
+            self._extrapolator = _Extrapolator1DNearest(x, f, extrapolation_range)
+        elif extrapolator_idx == EXT_LINEAR:
+            self._extrapolator = _Extrapolator1DLinear(x, f, extrapolation_range)
+        elif extrapolator_idx == EXT_QUADRATIC:
+            raise NotImplementedError("Quadratic extrapolator not implemented.")
+        else:
+            raise ValueError(f'Unsupported extrapolator type {extrapolation_type}.')
+
+        # create extrapolator per interpolation_type argument
+        if interpolator_idx == INT_LINEAR:
+            self._interpolator = _Interpolator1DLinear(self._x, self._f)
+        elif interpolator_idx == INT_CUBIC:
+            self._interpolator = _Interpolator1DCubic(self._x, self._f)
+        # elif interpolator_idx == INT_CUBIC_CONSTRAINED:
+        #     raise NotImplementedError("Constrained cubic interpolator not implemented.")
+        else:
+            raise ValueError(f'Interpolation type {interpolation_type} not supported.')
+
+    cdef double evaluate(self, double x) except? -1e999:
+        """
+        Evaluates the interpolating function.
+
+        :param double x: the point for which an interpolated value is required
+        :return: the interpolated value at point x.
+        """
+        cdef int index = find_index(self._x, x)
+
+        if index == -1:
+            if x < self._x[0] - self._extrapolation_range:
+                raise ValueError(
```

Comment:
Please write this on one line, it doesn't need to be split onto two. It would be perfectly legible on one.

Reply:
non essential style comment